Stable Diffusion 1.5 and Stable Diffusion XL Tutorials (#82)
triton-inference-server/tutorials: Popular_Models_Guide/StableDiffusion/README.md

[python based backend](https://github.com/triton-inference-server/backend/blob/main/docs/python_based_backends.md), [`/backend/diffusion/model.py`](backend/diffusion/model.py),
suitable for deploying multiple versions and configurations of
Diffusion models.

For more information on Stable Diffusion please visit
[stable-diffusion-v1-5](https://huggingface.co/runwayml/stable-diffusion-v1-5),
[stable-diffusion-xl](https://huggingface.co/docs/diffusers/en/using-diffusers/sdxl). For
more information on the TensorRT implementation please see the [TensorRT demo](https://github.com/NVIDIA/TensorRT/tree/release/9.2/demo/Diffusion).

> [!Note]
> This example is given as sample code and should be reviewed before use in production settings.

| [Requirements](#requirements) | [Building Server Image](#building-the-triton-inference-server-image) | [Stable Diffusion v1.5](#building-and-running-stable-diffusion-v-15) | [Stable Diffusion XL](#building-and-running-stable-diffusion-xl) | [Sending an Inference Request](#sending-an-inference-request) | [Model Configuration](docs/model_configuration.md) | [Client Application](docs/client_application.md) |

## Requirements

> The following instructions require a linux system with Docker

Reviewer: ```suggestion
The following instructions require a Linux system with Docker
```

installed. For CUDA support, make sure your CUDA driver meets the

> requirements in "NVIDIA Driver" section of [Deep Learning Framework

Reviewer: ```suggestion
requirements in the "NVIDIA Driver" section of [Deep Learning Framework
```

support matrix](https://docs.nvidia.com/deeplearning/frameworks/support-matrix/index.html).

## Building the Triton Inference Server Image

The example is designed based on the
`nvcr.io/nvidia/tritonserver:24.01-py3` docker image and [TensorRT OSS v9.2.0](https://github.com/NVIDIA/TensorRT/releases/tag/v9.2.0).

A set of convenience scripts are provided to create a docker image
based on the `nvcr.io/nvidia/tritonserver:24.01-py3` image with the
dependencies for the TensorRT Stable Diffusion demo installed.

> ### Trition Inference Server + TensorRT OSS

Reviewer 2: ```suggestion
### Triton Inference Server + TensorRT OSS
```

#### Clone Repository
```bash

> git clone https://github.com/triton-inference-server/tutorials.git -b nnshah1-stable-diffusion --single-branch

Reviewer 2: Should this branch be updated if merging to main?

cd tutorials/Popular_Models_Guide/StableDiffusion
```

#### Build `tritonserver:r24.01-diffusion` Image
```bash
./build.sh
```

#### Included Models

> The `default` build includes model configuration files for

Reviewer 3: ```suggestion
The `default` build includes model configuration files located in the `/diffusion-models` folder
```

[`stable_diffusion_1_5`](diffusion-models/stable_diffustion_1_5) and
[`stable_diffusion_xl`](diffusion-models/stable_diffustion_xl) but the
actual model artifacts and engine files are not included. They are
built in a separate step.

## Building and Running Stable Diffusion v 1.5

### Start `tritonserver:r24.01-diffusion` Container

The following command starts a container and volume mounts the current
directory as `workspace`.

```bash
./run.sh
```

### Build Stable Diffusion v 1.5 Engine

```bash

> ./scripts/build_models.sh --model stable_diffusion_1_5

Reviewer: I'm stepping through the tutorial and can't find this script in the container or repo folder. Is it possible it wasn't uploaded?

Author: thanks! uploaded.

```

#### Expected Output
```
 diffusion-models
|-- stable_diffusion_1_5
|   |-- 1
|   |   |-- 1.5-engine-batch-size-1
|   |   |-- 1.5-onnx
|   |   |-- 1.5-pytorch_model
|   `-- config.pbtxt

```

### Start a Server Instance

> [!Note]

> > We use `EXPLICIT` model control mode for demonstrtion purposes to control which stable diffusion version is loaded.

Reviewer: ```suggestion
> We use `EXPLICIT` model control mode for demonstration purposes to control which stable diffusion version is loaded.
```

> > We use `EXPLICIT` model control mode for demonstrtion purposes to control which stable diffusion version is loaded.

Reviewer 2: What're we demonstrating with explicit mode in this tutorial? I don't see load/unload calls being used, but might be missing it.

Reviewer 2: Is it just to avoid loading both models due to resource constraints if they had both models setup in the same model repository?

Author: because of build times and gpu memory wanted to avoid loading both at once - but wanted to keep both in the repo as samples. As an aside - having support for explicit load _only_ I think is useful :) - so no dynamic load but you can specify the set you want when launching ....

Reviewer 2: Agreed I think `--load-model` could be extended to default mode. Maybe a good new hire task? 

https://github.com/triton-inference-server/core/blob/d4fab30f3a946d242f3fad0f0690c511b916c282/src/model_repository_manager/model_repository_manager.cc#L422-L436

> > For production deployments please refer to [Secure Deploment Considerations](https://github.com/triton-inference-server/server/blob/main/docs/customization_guide/deploy.md) for more information on the risks associated with `EXPLICIT` mode.

Reviewer: ```suggestion
> For production deployments please refer to [Secure Deployment Considerations](https://github.com/triton-inference-server/server/blob/main/docs/customization_guide/deploy.md) for more information on the risks associated with `EXPLICIT` mode.
```

> > For production deployments please refer to [Secure Deployment Considerations](https://github.com/triton-inference-server/server/blob/main/docs/customization_guide/deploy.md) for more information on the risks associated with `EXPLICIT` mode.

Reviewer 3: nit: 80 char limit here

> ```baseh

Reviewer 2: ```suggestion
```bash
```

tritonserver --model-repository diffusion-models --model-control-mode explicit --load-model stable_diffusion_1_5
```

#### Expected Output
```
<SNIP>
I0229 20:15:52.125050 749 server.cc:676]
+----------------------+---------+--------+
| Model                | Version | Status |
+----------------------+---------+--------+
| stable_diffusion_1_5 | 1       | READY  |
+----------------------+---------+--------+

<SNIP>
```

## Building and Running Stable Diffusion XL

### Start `tritonserver:r24.01-diffusion` Container

The following command starts a container and volume mounts the current
directory as `workspace`.

```bash
./run.sh
```

### Build Stable Diffusion XL Engine

```bash
./scripts/build_models.sh --model stable_diffusion_xl
```

#### Expected Output
```
 diffusion-models
 |-- stable_diffusion_xl
    |-- 1
    |   |-- xl-1.0-engine-batch-size-1
    |   |-- xl-1.0-onnx
    |   `-- xl-1.0-pytorch_model
    `-- config.pbtxt
```

### Start a Server Instance

> [!Note]

> > We use `EXPLICIT` model control mode for demonstrtion purposes to control which stable diffusion version is loaded.

Reviewer: ```suggestion
> We use `EXPLICIT` model control mode for demonstration purposes to control which stable diffusion version is loaded.
```

> > For production deployments please refer to [Secure Deploment Considerations](https://github.com/triton-inference-server/server/blob/main/docs/customization_guide/deploy.md) for more information on the risks associated with `EXPLICIT` mode.

Reviewer: ```suggestion
> For production deployments please refer to [Secure Deployment Considerations](https://github.com/triton-inference-server/server/blob/main/docs/customization_guide/deploy.md) for more information on the risks associated with `EXPLICIT` mode.
```

> ```baseh

Reviewer 2: ```suggestion
```bash
```

## Requirements

The following instructions require a Linux system with Docker
installed. For CUDA support, make sure your CUDA driver meets the
requirements in the "NVIDIA Driver" section of [Deep Learning Framework
support matrix](https://docs.nvidia.com/deeplearning/frameworks/support-matrix/index.html).

## Building the Triton Inference Server Image

The example is designed based on the
`nvcr.io/nvidia/tritonserver:24.01-py3` docker image and [TensorRT OSS v9.2.0](https://github.com/NVIDIA/TensorRT/releases/tag/v9.2.0).

A set of convenience scripts are provided to create a docker image
based on the `nvcr.io/nvidia/tritonserver:24.01-py3` image with the
dependencies for the TensorRT Stable Diffusion demo installed.

### Triton Inference Server + TensorRT OSS

#### Clone Repository
```bash

> git clone https://github.com/triton-inference-server/tutorials.git -b r24.02 --single-branch
> cd tutorials/Popular_Models_Guide/StableDiffusion

Reviewer 2: I think saying "This is based on 24.01" and "git clone 24.02" may confuse people. Can we just use `main` branch for this clone?
```suggestion
git clone https://github.com/triton-inference-server/tutorials.git -b r24.02 --single-branch
cd tutorials/Popular_Models_Guide/StableDiffusion
```

```

#### Example Output

```bash
Client: 0 Throughput: 0.1825067711674996 Avg. Latency: 5.465569257736206
Throughput: 0.18224859609447058 Total Time: 5.487010717391968
```

If `--save-image` is given then output images will be saved as jpegs.

`
 client_0_generated_image_0.jpg
`

![sample_generated_image](./docs/client_0_generated_image_0_xl.jpg)

## Sample Client

The sample [client](client.py) application enables users to quickly
test the diffusion models under different concurrency scenarios. For a
full list and description of the client application's options use:

```
python3 client.py --help
```

### Sending Concurrent Requests

To increase load and concurrency users can use the `clients` and
`requests` options to control the number of client processes and the
number of requests sent by each client.

#### Example: Ten Clients Sending Ten Requests Each

The following command enables ten clients each sending ten
requests. Each client is an independent process that sends its
requests one after the other in parallel with the other nine clients.

```bash
python3 client.py --model stable_diffusion_xl --requests 10 --clients 10
```

## Known Issues and Limitations

> 1. When shutting down the server an invalid memory operation occurs:

Reviewer 2: Maybe some extension to this saying that it's harmless/expected

   ```
   free(): invalid pointer
   ```

   > [!Note]
   > This error is also seen in standalone applications outside of the Triton Inference Server
   > and we believe this is due to an interaction between imported python modules.

> 2. The diffusion backend doesn't support using a refiner model.

Reviewer 2: Link to an example refiner model/doc that came up to motivate this note?

full list and description of the client application's options use:

```
python3 client.py --help
```

### Sending Concurrent Requests

To increase load and concurrency users can use the `clients` and
`requests` options to control the number of client processes and the
number of requests sent by each client.

#### Example: Ten Clients Sending Ten Requests Each

The following command enables ten clients each sending ten
requests. Each client is an independent process that sends its
requests one after the other in parallel with the other nine clients.

```bash
python3 client.py --model stable_diffusion_xl --requests 10 --clients 10
```

## Known Issues and Limitations

1. When shutting down the server, an invalid memory operation occurs:

   > [!Note]
   > This error is also seen in standalone applications outside of the Triton Inference Server
   > and we believe this is due to an interaction between imported python modules. Further
   > we haven't seen any issues related to this error and believe it can be safely
   > ignored.

   ```
   free(): invalid pointer
   ```

2. The diffusion backend doesn't yet support using an optional refiner

>    model unlike the [demo][demo_reference] its based on. See also

Reviewer: ```suggestion
   model unlike the [demo][demo_reference] it's based on. See also
```

## Building the Triton Inference Server Image

The example is designed based on the
`nvcr.io/nvidia/tritonserver:YY.MM-py3` docker image and [TensorRT OSS v9.2.0](https://github.com/NVIDIA/TensorRT/releases/tag/v9.2.0).

A set of convenience scripts are provided to create a docker image
based on the `nvcr.io/nvidia/tritonserver:YY.MM-py3` image with the
dependencies for the TensorRT Stable Diffusion demo installed.

### Triton Inference Server + TensorRT OSS

#### Clone Repository
```bash

> git clone https://github.com/triton-inference-server/tutorials.git --single-branch

Reviewer 2: Didn't you want to lock the branch here for reproducibility in blog?

Author: I'll add the -b in the blog instructions - here we can leave it as main so it would work as we update it.
